### ISSUE TITLE
create test plan for Codex

### ISSUE THREAD

**Issue by jessiebroke** [status-im/nim-codex#345]

**Comment by c-blake** [comment]
There have been various discussions & proposals for client test plans.  Here is an initial sketch of various very basic scenarios we would like to work/study to make the conversation more concrete.

In all of these it would probably make sense to use files at least 5 GiB in size, and to test from at least 1 to 10 nodes. Also, in all of them baseline performance scaling should at least be measured, e.g. do 4 parallel downloads take the same time as 1 that is 4x as many bytes?  And similar adjustments to the dynamic changes.

  - A) Parallel Loads
    - Multiple parallel downloads
    - Multiple parallel uploads
    - Mixtures of parallel uploading & downloading
  - B) Loads while membership changes
    - Single downloads while nodes are being added
    - Single uploads while nodes are being added
    - Single downloads while nodes are being both added & dropped
    - Single uploads while nodes are being both added & dropped
  - C) Mixed Loads
    - All the A) scenarios while adding
    - All the A) scenarios while dropping
    - All the A) scenarios while both adding & dropping
  - D) Various network configurations
    - 1 node behind a NAT
    - All nodes behind NATs
    - All nodes behind multiple layers of NATs
    - Containerized/Virtual Machine Nodes
 
Others should please chime in with simple to describe but omitted ideas in this sketch!

**Comment by c-blake** [comment]
As an addendum to this - running test servers for more than the minimum few minutes and monitoring / logging used memory footprint would be a good "meta" property of all these tests.  A few hours or days would be ideal to be sure there are no leaks of file descriptors, memory, etc., etc.

Another meta-level property might be diverse distributions of host-to-host round-trip latencies & available bandwidths.  E.g., most measurements of real networks indicate they are in (or at least close to) the family of [stable distributions](https://en.wikipedia.org/wiki/Stable_distribution). Having these distributions be non-stationary (i.e. time varying mu/sigma/alpha/beta) would be an even higher level of realism.  (There are links for C/R/Python libs for inverse CDF/quantile/random deviates linked in that wiki.  Recording & re-playing real network delays & throughputs would be the ultimate in fidelity, of course, depending upon how scientific we regard as necessary to be informative.)

**Comment by benbierens** [comment]
I think this could turn out to be quite a sizable wishlist. It all does make sense to me, though. Since we're currently busy figuring out the infrastructure for these tests, I propose a few test scenario scripts are created. We can just write them in plain English and later on re-write them to whatever language we're testing it. I'm thinking have 2 or 3 simple but different enough scenarios written down will help us bring into focus the requirements for the infrastructure.

**Comment by c-blake** [comment]
I agree converting my wish list to a closer-to-pseudo-code format could add value.  Then that might be manually translated to HCL/Terraform or whatever other specification language backs some infrastructure-as-code solution.  Also, just to cross-link the issues I will mention the higher-level/project-level https://github.com/status-im/nim-codex/issues/350 here.
